pkg/agent/watch.go

```diff
@@ -201,60 +203,122 @@ func (e vmEvent) Format(state fmt.State, verb rune) {
+	addCPUMetric(vmResourceValueMax, vm.Spec.Guest.CPUs.Max)
+	addCPUMetric(vmResourceValueSpecUse, vm.Spec.Guest.CPUs.Use)
+	addCPUMetric(vmResourceValueStatusUse, vm.Status.CPUs)
+	return metrics
 }
 
-func setVMMetrics(vm *vmapi.VirtualMachine, nodeName string, perVMMetrics *PerVMMetrics) {
-	if vm.Status.Node != nodeName {
-		return
+func makeVMMemMetrics(vm *vmapi.VirtualMachine) []vmMetric {
+	var metrics []vmMetric
+
+	addMemMetric := func(resValType vmResourceValueType, val *int64) {
+		if val == nil {
+			return
+		}
+		labels := makeLabels(vm.Namespace, vm.Name, vm.Labels[endpointLabel], resValType)
+		metrics = append(metrics, vmMetric{
+			labels: labels,
+			value:  float64(*val),
+		})
 	}
 
-	commonLabelValues := []string{
-		vm.Namespace,
-		vm.Name,
-		vm.Labels[endpointLabel],
+	specMemVal := func(m *int32) *int64 {
+		if m == nil {
+			return nil
+		}
+		memValue := vm.Spec.Guest.MemorySlotSize.Value() * int64(*m)
+		return &memValue
+	}
+	statusMemVal := func(m *resource.Quantity) *int64 {
+		memValue := m.Value()
+		return &memValue
 	}
 
-	// CPU metrics derived from spec
-	specCPU := []kv[vmResourceValueType, *vmapi.MilliCPU]{
-		{key: vmResourceValueMin, value: vm.Spec.Guest.CPUs.Min},
-		{key: vmResourceValueMax, value: vm.Spec.Guest.CPUs.Max},
-		{key: vmResourceValueSpecUse, value: vm.Spec.Guest.CPUs.Use},
+	addMemMetric(vmResourceValueMin, specMemVal(vm.Spec.Guest.MemorySlots.Min))
+	addMemMetric(vmResourceValueMax, specMemVal(vm.Spec.Guest.MemorySlots.Max))
+	addMemMetric(vmResourceValueSpecUse, specMemVal(vm.Spec.Guest.MemorySlots.Use))
+	addMemMetric(vmResourceValueStatusUse, statusMemVal(vm.Status.MemorySize))
+	return metrics
+}
+
+func setVMMetrics(perVMMetrics *PerVMMetrics, vm *vmapi.VirtualMachine, nodeName string) {
+	if vm.Status.Node != nodeName {
+		return
 	}
-	for _, t := range specCPU {
-		if t.value != nil {
-			perVMMetrics.cpu.WithLabelValues(append(commonLabelValues, string(t.key))...).Set(t.value.AsFloat64())
-		}
+
+	cpuMetrics := makeVMCPUMetrics(vm)
+	for _, m := range cpuMetrics {
+		perVMMetrics.cpu.With(m.labels).Set(m.value)
 	}
 
-	if vm.Status.CPUs != nil {
-		perVMMetrics.cpu.WithLabelValues(append(commonLabelValues, string(vmResourceValueStatusUse))...).Set(vm.Status.CPUs.AsFloat64())
+	memMetrics := makeVMMemMetrics(vm)
+	for _, m := range memMetrics {
+		perVMMetrics.memory.With(m.labels).Set(m.value)
 	}
-	// Memory metrics derived from spec
-	specMem := []kv[vmResourceValueType, *int32]{
-		{key: vmResourceValueMin, value: vm.Spec.Guest.MemorySlots.Min},
-		{key: vmResourceValueMax, value: vm.Spec.Guest.MemorySlots.Max},
-		{key: vmResourceValueSpecUse, value: vm.Spec.Guest.MemorySlots.Use},
+}
+
+func updateVMMetrics(perVMMetrics *PerVMMetrics, oldVM, newVM *vmapi.VirtualMachine, nodeName string) {
+	if newVM.Status.Node != nodeName || oldVM.Status.Node != nodeName {
+		// The vm is either has been removed from the nodeName or has been added
+		// there.
```

Comment:
not sure this is correct - `updateVMMetrics` is called on all VMs, not just the ones on the autoscaler-agent's node, right?

Reply:
fixed in 5c55e03513341499fc4d5e0e1f4195291453025d.